user-service/src/main/java/rs/edu/raf/banka1/services/implementations/CapitalServiceImpl.java

```diff
@@ -36,4 +63,207 @@ public Capital createCapitalForListing(ListingType listingType, Long listingId,
+        capital.setTotal(capital.getTotal() + capital.getReserved());
+
+        if(bankAccount != null) {
+            bankAccount.setAvailableBalance(bankAccount.getAvailableBalance() + amount);
+        }
+
+        capitalRepository.save(capital);
+    }
+    private void processAddBalance(Capital capital, Double amount) {
+        if(amount <= 0)
+            throw new InvalidCapitalAmountException(amount);
+
+        BankAccount bankAccount = capital.getBankAccount();
+
+
+        capital.setTotal(capital.getTotal() + amount);
+
+        if(bankAccount != null) {
+            bankAccount.setBalance(bankAccount.getBalance() + amount);
+            bankAccount.setAvailableBalance(bankAccount.getAvailableBalance() + amount);
+        }
+
+        capitalRepository.save(capital);
+    }
+    private void processRemoveBalance(Capital capital, Double amount) {
+        BankAccount bankAccount = capital.getBankAccount();
+
+        if(amount <= 0)
+            throw new InvalidCapitalAmountException(amount);
+
+        double available = capital.getTotal() - capital.getReserved();
+
+        if(amount > available)
+            throw new NotEnoughCapitalAvailableException();
+
+        capital.setTotal(capital.getTotal() - amount);
+
+        if(bankAccount != null) {
+            if(bankAccount.getAvailableBalance() < amount) {
+                throw new NotEnoughCapitalAvailableException();
+            }
+            bankAccount.setAvailableBalance(bankAccount.getAvailableBalance() - amount);
+            bankAccount.setBalance(bankAccount.getBalance() - amount);
+        }
+
+        capitalRepository.save(capital);
+    }
+
+    public List<CapitalDto> getCapitalForListing(String accountNumber, ListingType listingType) {
+        BankAccount bankAccount = this.bankAccountRepository.findBankAccountByAccountNumber(accountNumber).orElseThrow(BankAccountNotFoundException::new);
+        return this.capitalRepository.getCapitalsByBankAccountAndListingType(bankAccount, listingType).stream().map(capitalMapper::capitalToCapitalDto).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<CapitalDto> getAllCapitals(String accountNumber) {
+        BankAccount bankAccount = this.bankAccountRepository.findBankAccountByAccountNumber(accountNumber).orElseThrow(BankAccountNotFoundException::new);
+        return this.capitalRepository.getCapitalsByBankAccount(bankAccount).stream().map(capitalMapper::capitalToCapitalDto).collect(Collectors.toList());
+    }
+
+    @Override
+    public Double estimateBalanceForex(String accountNumber, Long forexId) {
+        BankAccount bankAccount = this.bankAccountRepository.findBankAccountByAccountNumber(accountNumber).orElseThrow(BankAccountNotFoundException::new);
+        ListingForexDto listingForexDto = this.marketService.getForexById(forexId);
+        Capital capital = this.capitalRepository.getCapitalByListingIdAndBankAccount(forexId, bankAccount);
```

Comment:
Ovo uvek vraća praznu listu, jer kada imamo `bankAccount`, onda nemamo `listingId i obrnuto. Takođe, kada radite sa listingom u kapitalu, obavezno morate da pitate i za `listingType`, pošto su listinzi u različitim bazama, pa za jedan `listingId` može da vrati više stvari.